core/src/KokkosExp_MDRangePolicy.hpp

```diff
@@ -314,7 +315,42 @@ struct MDRangePolicy : public Kokkos::Impl::PolicyTraits<Properties...> {
   }
   bool impl_tune_tile_size() const { return m_tune_tile_size; }
 
+  tile_type tile_size_recommended() const {
+    tile_type rec_tile_sizes = {};
+
+    for (std::size_t i = 0; i < rec_tile_sizes.size(); ++i) {
+      rec_tile_sizes[i] = tile_size_recommended(i);
+    }
+    return rec_tile_sizes;
+  }
+
+  int tile_size_recommended(const int tile_rank) const {
```

Comment:
Do we actually need this API? Wouldn't it be sufficient for users to just query all recommended tile sizes and index into them if necessary?
In particular, I think it would be a good idea to actually use `tile_size_recommended` in the internal implementation so that we can be sure the user gets to see what we use.

Reply:
> Do we actually need this API? Wouldn't it be sufficient for users to just query all recommended tile sizes and index into them if necessary?

Yes, getting all recommend tile sizes should be enough, although it will require users to look up and know that the return type is a Kokkos::Array. I think this is ultimately a preference. For now, I will change the function that takes in a tile_rank to be private.

> In particular, I think it would be a good idea to actually use tile_size_recommended in the internal implementation so that we can be sure the user gets to see what we use.

There is a catch to this. The recommended tile size function simply looks up statically defined values and returns the tile sizes for all ranks, adjusted solely based on the product of default tile sizes.
In contrary, the internal implementation keeps track of the product of tile sizes which may include user specified tile sizes.

For instance, if `MDRangePolicy({dimensions ...}, {dimensions ...}, tile_type{{max_tile_size, 0, 1}}` is constructed, the internal implementation will set the tile size of rank 1 be `1` in order to avoid the product exceeding the max tile size. But the `recommended_tile_size()` will return `1` only if the product of default tile sizes exceeds the max tile size by rank 1.
Only partially applied. [90dd1a4](https://github.com/kokkos/kokkos/pull/6839/commits/90dd1a48353f140f7796cbcdf547a988e0df67a8).